### PR TITLE
flyway db migration tool plugin

### PR DESCRIPTION
**Pull request by senthilhns**
flyway db migration tool plugin plugin code 

Pipeline test cases for clean, migrate, baseline, validate and repair
https://app.harness.io/ng/account/gCoPSwHxS7ipOgx2iA9tOQ/all/orgs/default/projects/jenkinstoharness/pipelines/flyway_senthil_s01/executions/ISddRdssRUmvjvbqSmWcnQ/pipeline

Pipeline windows build
https://app.harness.io/ng/account/gCoPSwHxS7ipOgx2iA9tOQ/all/orgs/default/projects/jenkinstoharness/pipelines/flyway_senthil_win_03/executions/N4cY6fItSPmR-vG2op1PxQ/pipeline

p21

### CODE REVIEW

File: docker/Dockerfile.linux.arm
Comment (Ompragash) [inline docker/Dockerfile.linux.arm:1]: we don't need this and just arm64 should do

File: docker/Dockerfile.linux.arm64 [hunk 1]
```diff
@@ -0,0 +1,10 @@
+FROM alpine:3.6 as alpine
```
Comment (Ompragash) [inline docker/Dockerfile.linux.arm64]: update the base image version to 3.20 similar to the amd64 Dockerfile

File: docker/manifest.tmpl [hunk 2]
```diff
@@ -0,0 +1,31 @@
+image: harnesscommunity/drone-flyway-runner:{{#if build.tag}}{{trimPrefix "v" build.tag}}{{else}}latest{{/if}}
```
Comment (Ompragash) [inline docker/manifest.tmpl]: repo should be `plugins/flyway`

File: plugin/plugin.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+	}
+	return nil
+}
+
+func (p *FlywayPlugin) Run() error {
+	var stdoutBuf, stderrBuf bytes.Buffer
+	var err error
+
+	p.ExecCommand = p.GetExecArgsStr()
+	fmt.Println("Command: ", p.ExecCommand)
```
Comment (Ompragash) [inline plugin/plugin.go]:
```suggestion
	logrus.Infof("Executing command: %s", p.ExecCommand)
```

use logrus for consistency

File: plugin/plugin.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+	}
+
+	if p.InputArgs.IsDryRun == "TRUE" {
+		return nil
+	}
+
+	err = cmd.Run()
+	if err == nil {
+		fmt.Println(stdoutBuf.String())
+		fmt.Printf("Command execution success")
```
Comment (Ompragash) [inline plugin/plugin.go]:
```suggestion
		logrus.Infof("Command executed successfully: %s", stdoutBuf.String())
```

File: plugin/plugin.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+	} else {
+		fmt.Println(stderrBuf.String())
+		fmt.Printf("Error executing command: %v\n", err.Error())
```
Comment (Ompragash) [inline plugin/plugin.go]:
```suggestion
		logrus.Errorf("Command execution failed: %s, error: %v", stderrBuf.String(), err)
```

File: plugin/plugin.go [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+	ma := []mandatoryArg{
+		{"FLYWAY_URL", &args.Url, "url"},
+		{"FLYWAY_USER", &args.UserName, "username"},
+		{"FLYWAY_PASSWORD", &args.Password, "password"},
+	}
+
+	for _, m := range ma {
+		if os.Getenv(m.EnvName) == "" && *m.ParamName == "" {
+			LogPrintln("Missing mandatory argument: " + m.EnvName)
+			return fmt.Errorf("Missing mandatory argument: %s", m.Hint)
```
Comment (Ompragash) [inline plugin/plugin.go]:
```suggestion
			return fmt.Errorf("missing mandatory argument: %s (env: %s)", m.Hint, m.EnvName)
```

File: plugin/plugin.go [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+	return nil
+}
+
+var knownCommandsMap = map[string]bool{
+	MigrateCommand:  true,
+	CleanCommand:    true,
+	BaselineCommand: true,
+	RepairCommand:   true,
+	ValidateCommand: true,
+}
```
Comment (Ompragash) [inline plugin/plugin.go]: instead of map[string]bool use a map[string]string to associate commands with descriptions for clarity - eg: `MigrateCommand:  "Performs database migration",`

File: plugin/plugin.go [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+	} else {
+		fmt.Println(stderrBuf.String())
+		fmt.Printf("Error executing command: %v\n", err.Error())
+	}
+
+	return nil
+}
+
+func (p *FlywayPlugin) GetExecArgsStr() string {
+	var execCommand string
```
Comment (Ompragash) [inline plugin/plugin.go]: can we use `strings.Builder` instead?

File: plugin/plugin.go [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+	return nil
+}
+
+func (p *FlywayPlugin) IsUnknownCommand() error {
+	_, ok := knownCommandsMap[p.InputArgs.FlywayCommand]
+	if ok {
+		return nil
+	}
+
+	return fmt.Errorf("Unknown command: %s", p.InputArgs.FlywayCommand)
```
Comment (Ompragash) [inline plugin/plugin.go]:
```suggestion
        if _, ok := knownCommandsMap[p.InputArgs.FlywayCommand]; !ok {
            return fmt.Errorf("unknown command: %s", p.InputArgs.FlywayCommand)
        }
        return nil
```

File: plugin/plugin_test.go
Comment (Ompragash) [inline plugin/plugin_test.go:1]: add tests for edge cases
- missing required inputs (url, username, password)
- wrong Flyway command (invalidCommand)
- command with extra args (-X or -q)

File: plugin/plugin_test.go [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+		getDefaultPluginCommandLineArgs(),
+		getDefaultPluginUrl(),
+		getDefaultPluginUser(),
+		getDefaultPluginPassword(),
+	)
+
+	fp, err := Exec(context.TODO(), args)
+	if err != nil {
+		fmt.Println("Error in Exec: " + err.Error())
+		t.Fail()
```
Comment (Ompragash) [inline plugin/plugin_test.go]: can you replace t.Fail() statements with `require` and add failure messages as well

**Review by Ompragash — CHANGES_REQUESTED** [review]

File: plugin/plugin.go (hunk 4 above)
Comment (senthilhns) [reply plugin/plugin.go]: done

File: plugin/plugin.go (hunk 9 above)
Comment (senthilhns) [reply plugin/plugin.go]: done

File: plugin/plugin.go (hunk 8 above)
Comment (senthilhns) [reply plugin/plugin.go]: done

File: plugin/plugin.go (hunk 7 above)
Comment (senthilhns) [reply plugin/plugin.go]: done

File: plugin/plugin_test.go (hunk 10 above)
Comment (senthilhns) [reply plugin/plugin_test.go]: done

File: plugin/plugin.go (hunk 5 above)
Comment (senthilhns) [reply plugin/plugin.go]: done

File: plugin/plugin.go [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+
+	cmd := exec.Command(cmdName, cmdArgs...)
+	cmd.Stdout = &stdoutBuf
+	cmd.Stderr = &stderrBuf
+	if len(p.InputArgs.DriverPath) > 0 {
+		p.Env = "CLASSPATH=" + p.InputArgs.DriverPath
+		cmd.Env = append(os.Environ(), "CLASSPATH="+p.InputArgs.DriverPath)
+	}
+
+	if p.InputArgs.IsDryRun == "TRUE" {
```
Comment (Ompragash) [inline plugin/plugin.go]: use bool instead of string

File: plugin/plugin.go [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+	}
+	return nil
+}
+
+func (p *FlywayPlugin) Run() error {
+	var stdoutBuf, stderrBuf bytes.Buffer
+	var err error
+
+	p.ExecCommand = p.GetExecArgsStr()
+	logrus.Infof("Executing command: %s", p.ExecCommand)
```
Comment (Ompragash) [inline plugin/plugin.go]:
```suggestion
	logrus.Infof("Executing command: %s", strings.ReplaceAll(p.ExecCommand, p.InputArgs.Password, "******"))
```

File: plugin/plugin.go [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+func (p *FlywayPlugin) GetExecArgsStr() string {
+	var builder strings.Builder
+
+	builder.WriteString(GetFlywayExecutablePath() + " ")
+	builder.WriteString(p.InputArgs.FlywayCommand + " ")
+	builder.WriteString(p.CommandSpecificArgs + " ")
+
+	if len(p.InputArgs.Url) > 0 {
+		builder.WriteString("-url=" + p.InputArgs.Url + " ")
+	}
```
Comment (Ompragash) [inline plugin/plugin.go]:
```suggestion
	if p.InputArgs.Url != "" {
		builder.WriteString(fmt.Sprintf("-url=%s ", p.InputArgs.Url))
	}
```

File: plugin/plugin.go [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+		{"FLYWAY_USER", &args.UserName, "username"},
+		{"FLYWAY_PASSWORD", &args.Password, "password"},
+	}
+
+	for _, m := range ma {
+		if os.Getenv(m.EnvName) == "" && *m.ParamName == "" {
+			LogPrintln("Missing mandatory argument: " + m.EnvName)
+			return fmt.Errorf("Missing mandatory argument: %s (env: %s)", m.Hint, m.EnvName)
+		}
+	}
```
Comment (Ompragash) [inline plugin/plugin.go]:
```suggestion
	for _, m := range ma {
	        value := os.Getenv(m.EnvName)
		if value == "" && *m.ParamName == "" {
			eturn fmt.Errorf("missing mandatory argument: %s", m.Hint)
		}
		*m.ParamName = value
	}
```

File: plugin/plugin_test.go [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+		getDefaultPluginUrl(),
+		getDefaultPluginUser(),
+		getDefaultPluginPassword(),
+	)
+
+	fp, err := Exec(context.TODO(), args)
+	require.NoError(t, err, "Error in Exec: %v", err)
+
+	expectedCmd := " clean -cleanDisabled=false  -url=jdbc:mysql://3.4.9.2:3306/flyway_test " +
+		"-user=hnstest03 -password=sk89sl2@3 -locations=filesystem:/test/db-migrate01 "
```
Comment (Ompragash) [inline plugin/plugin_test.go]: replace hardcoded url, username and password with constants

```
const testUrl = "jdbc:mysql://mock.url:3306/test"
const testUser = "testUser"
const testPassword = "testPassword"
```

**Review by Ompragash — CHANGES_REQUESTED** [review]

File: plugin/plugin.go (hunk 11 above)
Comment (senthilhns) [reply plugin/plugin.go]: ok

File: plugin/plugin.go (hunk 14 above)
Comment (senthilhns) [reply plugin/plugin.go]: for _, m := range ma {
		if os.Getenv(m.EnvName) == "" && m.ParamName == "" {
			LogPrintln("Missing mandatory argument: " + m.EnvName)
			return fmt.Errorf("Missing mandatory argument: %s (env: %s)", m.Hint, m.EnvName)
		}
	} is just local check not adding *m.ParamName = value. Also struct mandatoryArg pointer type *ParamName to ParamName. Pointer may seem like that value is to be modified, but read only pass by value is sufficient

**Review by Ompragash — APPROVED** [review]
